File: android-nosql/src/main/java/com/github/florent37/androidnosql/AndroidNoSql.java
package com.github.florent37.androidnosql;

import android.content.Context;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Queue;
import java.util.concurrent.LinkedBlockingDeque;

import com.github.florent37.androidnosql.datasaver.DataSaver;
import com.github.florent37.androidnosql.datasaver.SharedPreferencesDataSaver;

/**
 * Created by florentchampigny on 29/05/2017.
 */

public class AndroidNoSql {
    private static final Queue<DataSaver> dataSavers = new LinkedBlockingDeque<>();

    public static Collection<DataSaver> getDataSaver() {
        return new ArrayList<>(dataSavers);
    }

    public static NoSql getInstance() {
        return NoSql.getInstance();
    }

    public static void initWith(DataSaver... savers) {
        for (DataSaver saver : savers) {
            addDataSaver(saver);
        }
    }

    public static void initWithDefault(Context context){
        initWith(new SharedPreferencesDataSaver(context));
    }

    public static void addDataSaver(DataSaver saver){
        synchronized (dataSavers) {
            final boolean wasEmpty = dataSavers.isEmpty();
            dataSavers.add(saver);
            if (wasEmpty && !dataSavers.isEmpty()) {
                final NoSql noSql = NoSql.getInstance();
                noSql.load();
            }
        }
    }

    public static void clearDataSavers(){
        for (DataSaver dataSaver : AndroidNoSql.getDataSaver()) {
            dataSaver.remove(NoSql.PATH_SEPARATOR);
        }
    }
}

File: android-nosql/src/main/java/com/github/florent37/androidnosql/AndroidNoSql.kt
package com.github.florent37.androidnosql

import android.content.Context
import com.github.florent37.androidnosql.datasaver.DataSaver
import com.github.florent37.androidnosql.datasaver.SharedPreferencesDataSaver
import java.util.*
import java.util.concurrent.LinkedBlockingDeque

object AndroidNoSql {
    private val dataSavers: Queue<DataSaver> = LinkedBlockingDeque()
    val dataSaver: Collection<DataSaver>
        get() = ArrayList(dataSavers)

    suspend fun initWith(vararg savers: DataSaver) {
        for (saver in savers) {
            addDataSaver(saver)
        }
    }

    suspend fun initWithDefault(context: Context?) {
        initWith(SharedPreferencesDataSaver(context!!))
    }

    suspend fun addDataSaver(saver: DataSaver) {
        val wasEmpty = dataSavers.isEmpty()
        dataSavers.add(saver)
        if (wasEmpty && !dataSavers.isEmpty()) {
            NoSqlKt.load()
        }
    }

    suspend fun clearDataSavers() {
        for (dataSaver in dataSaver) {
            dataSaver.remove(PATH_SEPARATOR)
        }
    }
}